### PR TITLE
M2: 12-bit packed sample unpacking

### LINKED ISSUE #11: 12-bit packed sample unpacking
Port the packed-mode path (`airspy_set_packing` support): 12-bit packed → 16-bit samples (`unpack_samples` in `airspy.c`).

- Bit-exact vs C; pure function, unit-tested with hand-built patterns covering alignment edges.
- Benchmark-friendly shape (this is a hot path).

Reference C source (local-only, gitignored): `original/airspyone_host`: `libairspy/src/airspy.c`

### PR DESCRIPTION
**Pull request by jasonherald**
Closes #11.

- `conversion::unpack_samples`: the packed-mode expander, bit expressions transcribed 1:1 from `unpack_samples` (airspy.c) — each 3-word group yields eight 12-bit samples. Hot-path shape: `chunks_exact` pairs, no allocation, no bounds panics.
- **Documented deviation / upstream bug found during the port**: C's consumer calls the unpacker with `sample_count = ((buffer_size/2)*4)/3 = 174762`, which is not a multiple of 8 — the final loop iteration reads 2 words past the 262144-byte input and writes ~5 `u16` past the output allocation (a real heap OOB in packed mode upstream). The port processes complete groups only and returns the sample count written; the pipeline consumer (#12) sizes accordingly (174760 vs C's overrun 174762).
- Crate-internal with a documented dead-code allow until the #12 pipeline consumes it (same pattern as the vendor-request layer before #8).

TDD: 5 tests written first and watched fail — single-group bit-layout parity via a test-only inverse packer, multi-group ordering, complete-group clamping on both sides, empty inputs, all-ones/all-zeros saturation (the all-ones group packs to three `0xFFFFFFFF` words, pinning the layout). 51 tests green, workspace clippy `-D warnings` all features, fmt clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added support for unpacking Airspy’s packed 12-bit sample data into usable sample values.
  - Processes complete data groups while safely ignoring incomplete input or output buffers.
  - Preserves sample ordering and supports the full 12-bit value range for reliable signal processing.

- **Tests**
  - Added coverage for bit-layout accuracy, multiple sample groups, partial buffers, empty data, and boundary values.

### CODE REVIEW

File: crates/libairspy/src/conversion.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+/// `sample_count` that is not a multiple of 8 (262144-byte buffers
+/// give 174762) and overruns both buffers on the final partial group
+/// — an upstream out-of-bounds this port does not reproduce.
+// Consumed by the sample-type pipeline (#12); until that lands the
+// unpacker has no callers outside its tests.
+#[allow(dead_code)]
+// Every assembled value is at most 12 bits ((0xFF << 4) | 0xF =
+// 0xFFF), but clippy cannot prove it through the shift-or pairs.
+#[allow(clippy::cast_possible_truncation)]
+pub(crate) fn unpack_samples(input: &[u32], output: &mut [u16]) -> usize {
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/libairspy/src/conversion.rs]: Adopted — unpack_samples now takes the raw &[u8] USB stream and loads words with u32::from_le_bytes: identical values to C's uint32_t* cast on its supported LE hosts, portable to BE, no alignment requirement, and it saves the #12 call site from a byte→word conversion (the bulk buffers are bytes).

File: crates/libairspy/src/conversion.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    let groups = core::cmp::min(
+        input.len() / WORDS_PER_GROUP,
+        output.len() / SAMPLES_PER_GROUP,
+    );
+    for (words, samples) in input
+        .chunks_exact(WORDS_PER_GROUP)
+        .zip(output.chunks_exact_mut(SAMPLES_PER_GROUP))
+        .take(groups)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/libairspy/src/conversion.rs]: Done — .take removed; zip's shorter-iterator termination is exactly the computed group count, which survives only for the return value.

File: crates/libairspy/src/conversion.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
+//! Sample-format conversion, starting with the 12-bit packed-mode
+//! unpacker (`unpack_samples` in airspy.c). The IQ converters land
+//! here with the DSP milestone.
+
+/// Words consumed per unpacked group (`i += 3` in C's loop).
+const WORDS_PER_GROUP: usize = 3;
+/// Samples produced per group (`j += 8`).
+const SAMPLES_PER_GROUP: usize = 8;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/libairspy/src/conversion.rs]: Partially adopted: BYTES_PER_GROUP/SAMPLES_PER_GROUP now cite unpack_samples (airspy.c) explicitly. The inline masks/shifts stay literal deliberately — for a bit-kernel transcription, the 1:1 literal expressions are what make the parity audit against the C loop trivial (a reviewer diffs them side-by-side); naming eight masks would add an indirection layer between the port and its specification. The transcription comment states this at the loop.

**Comment by jasonherald** [comment]
@coderabbitai Re the outside-diff fixture finding: added — a hard-coded 12-byte fixture (words 0x12345678 / 0x9ABCDEF0 / 0x13579BDF) with expected samples derived by hand from the C expressions, independent of pack_group, decoding to 0x123 0x456 0x789 0xABC 0xDEF 0x013 0x579 0xBDF.
